### PR TITLE
Add `MergeGitIgnoreRecipe` to merge existing .gitignore file with common entries

### LINKED ISSUE #532: Recipe for .gitignore update
### What feature do you want to see added?

Right now the .gitignore is just added when missing. 

We need some logic to merge with entries from archetype

### Upstream changes

None

### Are you interested in contributing this feature?

_No response_

### PR DESCRIPTION
**Pull request by nagu165**
Added the logic to merge the existing gitignore files with the archetype gitignore files

Fixes  #532

### Testing done
Yes

-> Added the MergeGitIgnoreRecipe to handle the merging of the existing gitignore files with the archetype gitignore files.
-> Added tests to verify that the recipe is working as expected.
-> Logic such that the archetype file can be provided as a constructor argument and the recipe takes care of merging.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by nagu165** [comment]
test was added only for the merging logic and not for the creation of gitignore file as it was mentioned that the creation of gitignore file already takes place if it didn't exist.so, I'm assuming it has already been taken care of

Thank you.

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    public String getDescription() {
+        return "Merges .gitignore entries from archetype with existing .gitignore file.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new GitIgnoreMerger(archetypeGitIgnorePath);
+    }
+
+    private static class GitIgnoreMerger extends MavenIsoVisitor<ExecutionContext> {
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java:39]: Why a MavenIsoVisitor since we are manipulating raw text file ?

A TreeVisitor looks more appropriate

By using Files you bypass changing the LST during recipe execution.

You can check how other recipes are generating new files or modifying them.

File: plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipeTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                *.iws
+                *.ipr
+                .idea
+                
+                # Eclipse project files
+                .settings
+                .classpath
+                .project
+                """,
+                sourceSpecs -> sourceSpecs.path("archetypes/common-files/.gitignore")
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipeTest.java]: This should not be part of the source files. I think the text file can be directly added as a constant on the recipe

File: plugin-modernizer-core/src/main/jte/pr-title-MergeGitIgnoreRecipe.jte [hunk 3]
```diff
@@ -0,0 +1,5 @@
+@import io.jenkins.tools.pluginmodernizer.core.model.Plugin
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/jte/pr-title-MergeGitIgnoreRecipe.jte:1]: Missing test on TemplateUtils to ensure the PR title matches we expect

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            try {
+                if (Files.exists(gitIgnorePath) && Files.exists(archetypeGitIgnorePath)) {
+                    String merged = mergeGitIgnoreFiles(
+                            Files.readString(gitIgnorePath), Files.readString(archetypeGitIgnorePath));
+                    Files.writeString(gitIgnorePath, merged);
+                }
+            } catch (IOException e) {
+                LOG.error("Error processing .gitignore files", e);
+            }
+            return document;
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java:62]: This literally ignore pretty much all errors. I don't think it's a good idea to not rethrow the exeption

**Review by jonesbusy — CHANGES_REQUESTED** [review]
Need a second iteration. The visitor is wrong since it's implement a MavenIsoVisitor suited for pom.xml manipulation.

Perhaps missing some test for the recipe (.gitignore not exists, no changes done, entries in different order etc....)

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java [hunk 5]
```diff
@@ -46,27 +46,35 @@ private static class GitIgnoreMerger extends TreeVisitor<PlainText, ExecutionCon
         @Override
         public PlainText visit(Tree tree, ExecutionContext ctx) {
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java:47]: In fact I found there is a more adapted visitor `PlainTextVisitor` more adapted than the TreeVisitor Did you found this as well ?

```java
PlainText visitText(PlainText text, ExecutionContext ctx) {
   ...
   return text.withText(mergedContent);
```

As I say OpenRewrite should not manipulate the files directly but it's LST. Then the LST is written back to source code

Comment (nagu165) [reply plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/recipes/MergeGitIgnoreRecipe.java:47]: > In fact I found there is a more adapted visitor `PlainTextVisitor` more adapted than the TreeVisitor Did you found this as well ?
> 
Yes, Sir I did find it but it wasn't used anywhere else in the project, so i was hesitant to use it.
> ```java
> PlainText visitText(PlainText text, ExecutionContext ctx) {
>    ...
>    return text.withText(mergedContent);
> ```
> 
> As I say OpenRewrite should not manipulate the files directly but it's LST. Then the LST is written back to source code

Understood sir.

**Comment by jonesbusy** [comment]
Thanks for your effort. I think we are good to go.

We can still improve in the future if we discover some edge case

**Review by jonesbusy — APPROVED** [review]

**Comment by nagu165** [comment]
> Thanks for your effort. I think we are good to go.
> 
I'm very grateful for your guidance sir, Thank you
> We can still improve in the future if we discover some edge case

Yes sir, I'd be happy to do so
Thank you.

**Comment by jonesbusy** [comment]
Test shouldMergeGitIgnoreEntries is badly implemented. I miss it. It creates two .gitignore files instead of checking content before/after.

**Comment by nagu165** [comment]
> Test shouldMergeGitIgnoreEntries is badly implemented. I miss it. It creates two .gitignore files instead of checking content before/after.

Yes sir, I get it now.
Thank you for pointing it out.
